quantize_tensor_per_channel ARM implementation

ajliu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46018 quantize_tensor_per_channel ARM implementation**
* #46017 Add benchmark for per channel tensor quantization

## Summary:
Currently on mobile devices quantize_tensor has a vectorized implementation using ARM intrinsics; however quantize_tensor_per_channel does not.

## Test Plan:
Build for ARM Neon
```
BUILD_MOBILE_BENCHMARK=1 BUILD_MOBILE_TEST=1 ANDROID_DEBUG_SYMBOLS=1 BUILD_PYTORCH_MOBILE=1 ANDROID_ABI="armeabi-v7a with NEON" ./scripts/build_android.sh  -DANDROID_CCACHE=$(which ccache) -DBUILD_BINARY=ON
```
Build for ARM64
```
BUILD_MOBILE_BENCHMARK=1 BUILD_MOBILE_TEST=1 ANDROID_DEBUG_SYMBOLS=1 BUILD_PYTORCH_MOBILE=1 ANDROID_ABI=arm64-v8a ./scripts/build_android.sh  -DANDROID_CCACHE=$(which ccache) -DBUILD_BINARY=ON
```
Run the test binary over adb shell with the commands below.
```
adb push build_android/bin/quantized_test /data/local/tmp
adb shell "/data/local/tmp/quantized_test"
```
Run the benchmark binary over adb shell with the commands below. 
```
adb push build_android/bin/quantize_per_channel /data/local/tmp/
adb shell "/data/local/tmp/quantize_per_channel"
```
Note that by android cpu is not frequency locked by default and can lead to noisy benchmark results, but this can be changed by running the following for every cpu.
```
adb shell "echo userspace > /sys/devices/system/cpu/${cpu}/cpufreq/scaling_governor"
adb shell "echo '2000000' > /sys/devices/system/cpu/${cpu}/cpufreq/scaling_setspeed"
```

Resulting benchmarks are located [here](https://gist.github.com/AJLiu/d1711bb6a5e93b3338eca2c14c8aec9f)
Google spreadsheet comparing results [here](https://docs.google.com/spreadsheets/d/1Ky-rEu2CqOqex2a84b67hB1VLAlfEDgAN2ZXe8IlGF8/edit?usp=sharing)

**Overall results:**
- aarch64
    - 2x slowdown on 2d tensor benchmarks
    - 4x speed up on 4d contiguous tensor benchmarks
    - 4x speed up on 4d channels last tensor benchmarks
- neon
    - 2x speed up on 2d tensor benchmarks
    - 20x speed up on 4d contiguous tensor benchmarks
    - 20x speed up on 4d contiguous tensor benchmarks

I suspect that this is an overall performance boost, however it adds a small amount of overhead that becomes very noticeable when quantizing small tensors.

Reviewers: kimishpatel

Subscribers:

Tasks: T76832258

Tags:

Differential Revision: [D24286528](https://our.internmc.facebook.com/intern/diff/D24286528)

ajliu committed:
quantize_tensor_per_channel ARM implementation

Summary: Currently on mobile devices quantize_tensor has a vectorized implementation
using ARM intrinsics; however quantize_tensor_per_channel does not.

Test Plan: Build and push to mobile device
```
BUILD_MOBILE_BENCHMARK=1 BUILD_MOBILE_TEST=1 ANDROID_DEBUG_SYMBOLS=1 BUILD_PYTORCH_MOBILE=1 ANDROID_ABI=arm64-v8a ./scripts/build_android.sh  -DANDROID_CCACHE=$(which ccache) -DBUILD_BINARY=ON
adb push build_android/bin/quantize_per_channel /data/local/tmp
```
and then run the benchmark binary over adb shell

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

AJLiu commented:
Closing because after some testing it appears that the compiler is performing these optimizations automatically.

ajliu committed:
Update on "quantize_tensor_per_channel ARM implementation"


Summary: Currently on mobile devices quantize_tensor has a vectorized implementation
using ARM intrinsics; however quantize_tensor_per_channel does not.

Test Plan: Build and push to mobile device
```
BUILD_MOBILE_BENCHMARK=1 BUILD_MOBILE_TEST=1 ANDROID_DEBUG_SYMBOLS=1 BUILD_PYTORCH_MOBILE=1 ANDROID_ABI=arm64-v8a ./scripts/build_android.sh  -DANDROID_CCACHE=$(which ccache) -DBUILD_BINARY=ON
adb push build_android/bin/quantize_per_channel /data/local/tmp
```
and then run the benchmark binary over adb shell

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ajliu committed:
Update on "quantize_tensor_per_channel ARM implementation"


## Summary:
Currently on mobile devices quantize_tensor has a vectorized implementation using ARM intrinsics; however quantize_tensor_per_channel does not.

## Test Plan:
Build for ARM Neon
```
BUILD_MOBILE_BENCHMARK=1 BUILD_MOBILE_TEST=1 ANDROID_DEBUG_SYMBOLS=1 BUILD_PYTORCH_MOBILE=1 ANDROID_ABI="armeabi-v7a with NEON" ./scripts/build_android.sh  -DANDROID_CCACHE=$(which ccache) -DBUILD_BINARY=ON
```
Build for ARM64
```
BUILD_MOBILE_BENCHMARK=1 BUILD_MOBILE_TEST=1 ANDROID_DEBUG_SYMBOLS=1 BUILD_PYTORCH_MOBILE=1 ANDROID_ABI=arm64-v8a ./scripts/build_android.sh  -DANDROID_CCACHE=$(which ccache) -DBUILD_BINARY=ON
```
Then run the benchmark binary over adb shell. Note that by android cpu is not frequency locked by default and can lead to noisy benchmark results, but this can be changed by running the following for every cpu.
```
adb shell "echo userspace > /sys/devices/system/cpu/${cpu}/cpufreq/scaling_governor"
adb shell "echo '2000000' > /sys/devices/system/cpu/${cpu}/cpufreq/scaling_setspeed"
adb push build_android/bin/quantize_per_channel /data/local/tmp/
adb shell "/data/local/tmp/quantize_per_channel"
```

Resulting benchmarks are located [here](https://gist.github.com/AJLiu/d1711bb6a5e93b3338eca2c14c8aec9f)
Google spreadsheet comparing results [here](https://docs.google.com/spreadsheets/d/1Ky-rEu2CqOqex2a84b67hB1VLAlfEDgAN2ZXe8IlGF8/edit?usp=sharing)

**Overall results:**
- aarch64
    - 2x slowdown on 2d tensor benchmarks
    - 4x speed up on 4d contiguous tensor benchmarks
    - 4x speed up on 4d channels last tensor benchmarks
- neon
    - 2x speed up on 2d tensor benchmarks
    - 20x speed up on 4d contiguous tensor benchmarks
    - 20x speed up on 4d contiguous tensor benchmarks

I suspect that this is an overall performance boost, however it adds a small amount of overhead that becomes very noticeable when quantizing small tensors.

Reviewers: kimishpatel

Subscribers:

Tasks: T76832258

Tags:

Differential Revision: [D24286528](https://our.internmc.facebook.com/intern/diff/D24286528)

[ghstack-poisoned]

AJLiu commented:
Latest commit minimizes the overhead slightly. Now when compiling with Neon we see an overall improvement compared to before, but Arm64 still has a bit of a slowdown on small tensors.

New benchmark outputs: https://gist.github.com/AJLiu/814232ac153b8d8029a8d37dc074320c

kimishpatel commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -2568,6 +2568,225 @@ void dequantize_tensor_per_tensor_affine_cpu(
 // TODO: add fbgemm for per channel
+#if defined(__ARM_NEON__) || defined(__aarch64__)
+// Generic template defaults to naive quantize implementation
+template <typename T>
+void quantize_tensor_per_channel_arm(
+    const float* in,
+    Tensor qtensor,
+    const double* scales,
+    const int64_t* zero_points,
+    const bool channels_last,
+    const int64_t batches,
+    const int64_t elements_per_channel,
+    const int64_t channels) {
+  auto out = qtensor.data_ptr<T>();
+  if (channels_last) {
+    for (auto b = 0; b < batches; ++b) {
+      for (auto e = 0; e < elements_per_channel; ++e) {
+        for (auto c = 0; c < channels; ++c) {
+          auto i = b * channels * elements_per_channel + e * channels + c;
+          out[i] =
+              at::native::quantize_val<T>(scales[c], zero_points[c], in[i]);
+        }
+      }
+    }
+  } else {
+    for (auto b = 0; b < batches; ++b) {
+      for (auto c = 0; c < channels; ++c) {
+        for (auto e = 0; e < elements_per_channel; ++e) {
+          auto i = b * channels * elements_per_channel +
+              c * elements_per_channel + e;
+          out[i] =
+              at::native::quantize_val<T>(scales[c], zero_points[c], in[i]);
+        }
+      }
+    }
+  }
+}
```
It seems that this implementation is same as fallback path here: https://github.com/pytorch/pytorch/pull/46018/files#diff-ecdf2908b94623646b614d155e2dd4997d657070cdf4846057f9dc49639bbbcfR2832. Can you move this function out of arm specific ifdef and make it usable for the fallback path as well.

Make sure that the comments in the fallback path are reflected here as well. Specifically channels_last condition. It is little confusion that channels last condition captures which axis is to be quantized.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -2568,6 +2568,225 @@ void dequantize_tensor_per_tensor_affine_cpu(
+  const int32x4_t vmagic_int = vdupq_n_s32(0x4B400000);
+  // Copy zero_points (int64_t) into int32_t array
+  // Copy reciprocal of scales (double) into float array
+  int32_t zero_points_int32t[channels];
```
Why not use std::vector?

kimishpatel replied:
I am honestly not familiar this way of allocating dynamic array.

AJLiu replied:
This is a static array, since we know that zero points should have exactly `channels` number of elements.

kimishpatel replied:
Yes but channels is not known at compile time, right?

AJLiu replied:
Hmm looked into it and it looks like the runtime difference won't really be any different. `std::vector` might use a little more memory, but probably negligible enough to just pick whichever is easier to read.

I'll switch over to using `std::vector`, since the current route above requires you to free the memory pointer manually.

kimishpatel replied:
That would be a very strong reason to do it.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -2568,6 +2568,225 @@ void dequantize_tensor_per_tensor_affine_cpu(
+        uint32_t c = 0;
+        while (c + 8 < channels) {
+          const int32x4_t voffset0123 =
+              vsubq_s32(vld1q_s32(zero_points_int32t + c), vmagic_int);
```
You can further optimize this if by folding this vmagic_int subtraction in the zero_points_int32t that you are copying over at the beginning.
Applies in non channels last case as well.

AJLiu replied:
good catch!

kimishpatel reviewed (changes requested):
Left a few comments.
Also can you add this test https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/test/quantized_test.cpp to mobile tests and run it through to make sure it passes.

ajliu committed:
Update on "quantize_tensor_per_channel ARM implementation"


## Summary:
Currently on mobile devices quantize_tensor has a vectorized implementation using ARM intrinsics; however quantize_tensor_per_channel does not.

## Test Plan:
Build for ARM Neon
```
BUILD_MOBILE_BENCHMARK=1 BUILD_MOBILE_TEST=1 ANDROID_DEBUG_SYMBOLS=1 BUILD_PYTORCH_MOBILE=1 ANDROID_ABI="armeabi-v7a with NEON" ./scripts/build_android.sh  -DANDROID_CCACHE=$(which ccache) -DBUILD_BINARY=ON
```
Build for ARM64
```
BUILD_MOBILE_BENCHMARK=1 BUILD_MOBILE_TEST=1 ANDROID_DEBUG_SYMBOLS=1 BUILD_PYTORCH_MOBILE=1 ANDROID_ABI=arm64-v8a ./scripts/build_android.sh  -DANDROID_CCACHE=$(which ccache) -DBUILD_BINARY=ON
```
Then run the benchmark binary over adb shell. Note that by android cpu is not frequency locked by default and can lead to noisy benchmark results, but this can be changed by running the following for every cpu.
```
adb shell "echo userspace > /sys/devices/system/cpu/${cpu}/cpufreq/scaling_governor"
adb shell "echo '2000000' > /sys/devices/system/cpu/${cpu}/cpufreq/scaling_setspeed"
adb push build_android/bin/quantize_per_channel /data/local/tmp/
adb shell "/data/local/tmp/quantize_per_channel"
```

Resulting benchmarks are located [here](https://gist.github.com/AJLiu/d1711bb6a5e93b3338eca2c14c8aec9f)
Google spreadsheet comparing results [here](https://docs.google.com/spreadsheets/d/1Ky-rEu2CqOqex2a84b67hB1VLAlfEDgAN2ZXe8IlGF8/edit?usp=sharing)

**Overall results:**
- aarch64
    - 2x slowdown on 2d tensor benchmarks
    - 4x speed up on 4d contiguous tensor benchmarks
    - 4x speed up on 4d channels last tensor benchmarks
- neon
    - 2x speed up on 2d tensor benchmarks
    - 20x speed up on 4d contiguous tensor benchmarks
    - 20x speed up on 4d contiguous tensor benchmarks

I suspect that this is an overall performance boost, however it adds a small amount of overhead that becomes very noticeable when quantizing small tensors.

Reviewers: kimishpatel

Subscribers:

Tasks: T76832258

Tags:

Differential Revision: [D24286528](https://our.internmc.facebook.com/intern/diff/D24286528)

[ghstack-poisoned]

ajliu committed:
Update on "quantize_tensor_per_channel ARM implementation"


## Summary:
Currently on mobile devices quantize_tensor has a vectorized implementation using ARM intrinsics; however quantize_tensor_per_channel does not.

## Test Plan:
Build for ARM Neon
```
BUILD_MOBILE_BENCHMARK=1 BUILD_MOBILE_TEST=1 ANDROID_DEBUG_SYMBOLS=1 BUILD_PYTORCH_MOBILE=1 ANDROID_ABI="armeabi-v7a with NEON" ./scripts/build_android.sh  -DANDROID_CCACHE=$(which ccache) -DBUILD_BINARY=ON
```
Build for ARM64
```
BUILD_MOBILE_BENCHMARK=1 BUILD_MOBILE_TEST=1 ANDROID_DEBUG_SYMBOLS=1 BUILD_PYTORCH_MOBILE=1 ANDROID_ABI=arm64-v8a ./scripts/build_android.sh  -DANDROID_CCACHE=$(which ccache) -DBUILD_BINARY=ON
```
Then run the benchmark binary over adb shell. Note that by android cpu is not frequency locked by default and can lead to noisy benchmark results, but this can be changed by running the following for every cpu.
```
adb shell "echo userspace > /sys/devices/system/cpu/${cpu}/cpufreq/scaling_governor"
adb shell "echo '2000000' > /sys/devices/system/cpu/${cpu}/cpufreq/scaling_setspeed"
adb push build_android/bin/quantize_per_channel /data/local/tmp/
adb shell "/data/local/tmp/quantize_per_channel"
```

Resulting benchmarks are located [here](https://gist.github.com/AJLiu/d1711bb6a5e93b3338eca2c14c8aec9f)
Google spreadsheet comparing results [here](https://docs.google.com/spreadsheets/d/1Ky-rEu2CqOqex2a84b67hB1VLAlfEDgAN2ZXe8IlGF8/edit?usp=sharing)

**Overall results:**
- aarch64
    - 2x slowdown on 2d tensor benchmarks
    - 4x speed up on 4d contiguous tensor benchmarks
    - 4x speed up on 4d channels last tensor benchmarks
- neon
    - 2x speed up on 2d tensor benchmarks
    - 20x speed up on 4d contiguous tensor benchmarks
    - 20x speed up on 4d contiguous tensor benchmarks

I suspect that this is an overall performance boost, however it adds a small amount of overhead that becomes very noticeable when quantizing small tensors.

Reviewers: kimishpatel

Subscribers:

Tasks: T76832258

Tags:

Differential Revision: [D24286528](https://our.internmc.facebook.com/intern/diff/D24286528)

[ghstack-poisoned]

AJLiu commented:
Addressed requested changes and enabled `pytorch/aten/src/ATen/test/quantized_test.cpp` for mobile.

However it looks like none of the tests in that file actually check quantize_tensor_per_channel so I added two new unit tests for them.

kimishpatel commented:
> Addressed requested changes and enabled `pytorch/aten/src/ATen/test/quantized_test.cpp` for mobile.
> 
> However it looks like none of the tests in that file actually test quantize_tensor_per_channel so I added two new unit tests for them (to check contiguous and channels_last formats). It looks like both new tests pass for both mobile architectures.

Thanks for these changes. Looks good overall. I am little perplexed as to why 2D tensors are so slow. Only thing different seems to be new allocations for inv scale and offset and writing those values out which maybe is adding extra memory overhead.

ajliu committed:
Update on "quantize_tensor_per_channel ARM implementation"


## Summary:
Currently on mobile devices quantize_tensor has a vectorized implementation using ARM intrinsics; however quantize_tensor_per_channel does not.

## Test Plan:
Build for ARM Neon
```
BUILD_MOBILE_BENCHMARK=1 BUILD_MOBILE_TEST=1 ANDROID_DEBUG_SYMBOLS=1 BUILD_PYTORCH_MOBILE=1 ANDROID_ABI="armeabi-v7a with NEON" ./scripts/build_android.sh  -DANDROID_CCACHE=$(which ccache) -DBUILD_BINARY=ON
```
Build for ARM64
```
BUILD_MOBILE_BENCHMARK=1 BUILD_MOBILE_TEST=1 ANDROID_DEBUG_SYMBOLS=1 BUILD_PYTORCH_MOBILE=1 ANDROID_ABI=arm64-v8a ./scripts/build_android.sh  -DANDROID_CCACHE=$(which ccache) -DBUILD_BINARY=ON
```
Run the test binary over adb shell with the commands below.
```
adb push build_android/bin/quantized_test /data/local/tmp
adb shell "/data/local/tmp/quantized_test"
```
Run the benchmark binary over adb shell with the commands below. 
```
adb push build_android/bin/quantize_per_channel /data/local/tmp/
adb shell "/data/local/tmp/quantize_per_channel"
```
Note that by android cpu is not frequency locked by default and can lead to noisy benchmark results, but this can be changed by running the following for every cpu.
```
adb shell "echo userspace > /sys/devices/system/cpu/${cpu}/cpufreq/scaling_governor"
adb shell "echo '2000000' > /sys/devices/system/cpu/${cpu}/cpufreq/scaling_setspeed"
```

Resulting benchmarks are located [here](https://gist.github.com/AJLiu/d1711bb6a5e93b3338eca2c14c8aec9f)
Google spreadsheet comparing results [here](https://docs.google.com/spreadsheets/d/1Ky-rEu2CqOqex2a84b67hB1VLAlfEDgAN2ZXe8IlGF8/edit?usp=sharing)

**Overall results:**
- aarch64
    - 2x slowdown on 2d tensor benchmarks
    - 4x speed up on 4d contiguous tensor benchmarks
    - 4x speed up on 4d channels last tensor benchmarks
- neon
    - 2x speed up on 2d tensor benchmarks
    - 20x speed up on 4d contiguous tensor benchmarks
    - 20x speed up on 4d contiguous tensor benchmarks

I suspect that this is an overall performance boost, however it adds a small amount of overhead that becomes very noticeable when quantizing small tensors.

Reviewers: kimishpatel

Subscribers:

Tasks: T76832258

Tags:

Differential Revision: [D24286528](https://our.internmc.facebook.com/intern/diff/D24286528)

[ghstack-poisoned]

kimishpatel reviewed (approved):
Looks good. Make sure CI is clean. I see some lint errors.

AJLiu commented:
Reran the benchmarks and the overhead is gone.
Overall results:
- aarch64
  - 1.4x speed up on 2d tensor benchmarks
  - 4x speed up on 4d contiguous tensor benchmarks
  - 4x speed up on 4d channels last tensor benchmarks
- neon
  - 3x speed up on 2d tensor benchmarks
  - 20x speed up on 4d contiguous tensor benchmarks
  - 20x speed up on 4d contiguous tensor benchmarks

AJLiu commented:
Lint errors are unrelated to this pr, looks like https://github.com/pytorch/pytorch/commit/b3eb0c86cf21d8dad5744a917c70d846a8715e69 fixes it

kimishpatel commented:
Sounds good.
